src/weights/data_type_weights.cpp

```diff
@@ -145,38 +146,62 @@ void data_type_weights<TensorDataType>::set_dims(std::vector<int> matrix_height_
 // -----------------------------------------------
 
 template <typename TensorDataType>
-data_type_weights_initializer<TensorDataType>* data_type_weights<TensorDataType>::get_initializer() {
-  return const_cast<data_type_weights_initializer<TensorDataType>*>(static_cast<const data_type_weights&>(*this).get_initializer());
+auto data_type_weights<TensorDataType>::get_initializer() -> InitializerType* {
+  return const_cast<InitializerType*>(static_cast<const data_type_weights&>(*this).get_initializer());
 }
 template <typename TensorDataType>
-const data_type_weights_initializer<TensorDataType>* data_type_weights<TensorDataType>::get_initializer() const {
+auto data_type_weights<TensorDataType>::get_initializer() const
+  -> const InitializerType* {
   return m_initializer.get();
 }
 template <typename TensorDataType>
-void data_type_weights<TensorDataType>::set_initializer(std::unique_ptr<weights_initializer>&& init) {
-  m_initializer = std::move(reinterpret_cast<std::unique_ptr<data_type_weights_initializer<TensorDataType>>&&>(init));
```

Comment:
`reinterpret_cast` should be avoided with class types. In the scope of this PR, this is unlikely to cause harm. However, in a world in which we might have multiple `weights_initializer`/`optimizer` types in a model, this would have been very dangerous.